File: kotlin/src/main/kotlin/com/peshchuk/topjslibs/GoogleSearchResult.kt
package com.peshchuk.topjslibs

private val SEARCH_RESULT_LINK =
    """<a href="([^"]+?)"[^>]+?data-ved="[^"]+?"[^>]+?onmousedown="[^"]+?"><br>""".toRegex()

class GoogleSearchResult(private val linksLimit: Int, private val fetchHtml: () -> String) {

    fun getLinks(): List<String> {
        val html = fetchHtml()
        val matchResults = SEARCH_RESULT_LINK.findAll(html)

        return matchResults
            .take(linksLimit)
            .filter { matchResult -> matchResult.groups.size > 1 }
            .map { matchResult -> matchResult.groups[1]?.value!! }
            .toList()
    }
}

File: scala/src/main/scala/com/peshchuk/topjslibs/GoogleSearchResult.scala
package com.peshchuk.topjslibs

class GoogleSearchResult(private val linksLimit: Int, private val fetchHtml: () => String) {

  def getLinks: List[String] = {
    val html = fetchHtml()
    val matchData = GoogleSearchResult.SEARCH_RESULT_LINK.findAllIn(html).matchData

    matchData
      .filter(_.groupCount > 0)
      .map(_.group(1))
      .slice(0, linksLimit)
      .toList
  }
}

object GoogleSearchResult {
  private final val SEARCH_RESULT_LINK =
    """<a href="([^"]+?)"[^>]+?data-ved="[^"]+?"[^>]+?onmousedown="[^"]+?"><br>""".r
}
